CBinding/Gui/CTextEditorExtension.cs

```diff
@@ -221,382 +245,274 @@ public override bool KeyPress (KeyDescriptor descriptor)
+		/// <param name="token">Token.</param>
 		public override Task<ICompletionDataList> HandleCodeCompletionAsync (CodeCompletionContext completionContext, char completionChar, CancellationToken token = default(CancellationToken))
 		{
-			string lineText = Editor.GetLineText (completionContext.TriggerLine).TrimEnd();
-			
-			// If the line ends with a matched extension, invoke its handler
-			foreach (KeyValuePair<string, GetMembersForExtension> pair in completionExtensions) {
-				if (lineText.EndsWith(pair.Key)) {
-					lineText = lineText.Substring (0, lineText.Length - pair.Key.Length);
-					
-					int nameStart = lineText.LastIndexOfAny (allowedCharsMinusColon) + 1;
-					string itemName = lineText.Substring (nameStart).Trim ();
-					
-					if (string.IsNullOrEmpty (itemName))
-						return null;
-					
-					return Task.FromResult ((ICompletionDataList)pair.Value (this, pair.Key, itemName));
+			ICompletionDataList list = new CompletionDataList ();
+			if (shouldCompleteOn(completionChar)) {
+				CProject project = DocumentContext.Project as CProject;
+				unsavedFiles = new List<CXUnsavedFile> ();
+				foreach (Document openDocument in IdeApp.Workbench.Documents) {
+					if (openDocument.IsDirty) {
+						CXUnsavedFile unsavedFile = new CXUnsavedFile ();
+						unsavedFile.Filename = openDocument.FileName;
+						unsavedFile.Length = openDocument.Editor.Text.Length;
+						unsavedFile.Contents = openDocument.Editor.Text;
+						if (project.BOMPresentInFile [openDocument.FileName]) {
+							unsavedFile.Length += 3;
+							unsavedFile.Contents = "   " + unsavedFile.Contents;
+						}
+						unsavedFiles.Add (unsavedFile);
+					}
 				}
+				string operatorPattern = "operator\\s*(\\+|\\-|\\*|\\/|\\%|\\^|\\&|\\||\\~|\\!|\\=|\\<|\\>|\\(\\s*\\)|\\[\\s*\\]|new|delete)";
+				bool fieldOrMethodMode = completionChar == '.' || completionChar == '>' ? true : false;
+				IntPtr pResults = project.cLangManager.codeComplete (completionContext, unsavedFiles.ToArray (), DocumentContext.Name);
+				CXCodeCompleteResults results = Marshal.PtrToStructure<CXCodeCompleteResults> (pResults);
+				if (results.Results.ToInt64 () != 0) {
+					for(int i = 0; i < results.NumResults; i++) {
+						IntPtr iteratingPointer = results.Results + i * Marshal.SizeOf<CXCompletionResult>();
+						CXCompletionResult resultItem = Marshal.PtrToStructure<CXCompletionResult> (iteratingPointer);
+						CXCompletionString completionString = new CXCompletionString(resultItem.CompletionString);
+						uint completionchunknum = clang.getNumCompletionChunks (completionString.Pointer);
+						for (uint j = 0; j < completionchunknum; j++) {
+							if (clang.getCompletionChunkKind(completionString.Pointer, j) != CXCompletionChunkKind.CXCompletionChunk_TypedText)
+								continue;
+							uint priority = clang.getCompletionPriority (completionString.Pointer);
+							CXString cxstring = clang.getCompletionChunkText(completionString.Pointer, j);
```

Comment:
This code is very dense and difficult to read, it would be better with some helpers (and empty lines).

For example:

    IEnumerable<ClangCompletionUnit> GetCompletionUnits (CXCompletionResult result)
    {
        var completionString = new CXCompletionString (result.CompletionString);
        uint chunkCount = clang.getNumCompletionChunks (completionString.Pointer);

        for (uint j = 0; j < chunkCount ; j++) {
            if (clang.getCompletionChunkKind (completionString.Pointer, j) != CXCompletionChunkKind.TypedText)
                continue;

            switch (resultItem.CursorKind) {
                case CXCursorKind.Destructor:
                case CXCursorKind.UnaryOperator:
                case CXCursorKind.BinaryOperator:
                case CXCursorKind.CompoundAssignOperator:
                    continue;
                case CXCursorKind.ClassDecl:
                case CXCursorKind.StructDecl:
                    if (fieldOrMethodMode)
                         continue;
                    break;
            }
            string completionText = ConvertString (clang.getCompletionChunkText(completionString.Pointer, j));

            if (operatorRegex.IsMatch (completionText))
                continue;

            uint priority = clang.getCompletionPriority (completionString.Pointer);
            yield return new ClangCompletionUnit (result, completionText , priority);
    }
    
    string ConvertString (CXString cxstring)
    {
        string s = Marshal.PtrToStringAnsi (clang.getCString (cxstring));
        clang.disposeString (cxstring);
        return s;
    }

Note that I've also moved operations around for efficiency so that for example we skip marshalling the completion text when we're skipping it due to the cursor type.